rust/lance-index/src/scalar/geo/bkd.rs

```diff
@@ -0,0 +1,882 @@
+//! - Stores tree structure separately from leaf data for lazy loading
+
+use arrow_array::cast::AsArray;
+use arrow_array::{ArrayRef, Float64Array, RecordBatch, UInt64Array};
+use arrow_schema::{DataType, Field, Schema};
+use deepsize::DeepSizeOf;
+use lance_core::{Result, ROW_ID};
+use snafu::location;
+use std::sync::Arc;
+
+// Schema field names
+const NODE_ID: &str = "node_id";
+const MIN_X: &str = "min_x";
+const MIN_Y: &str = "min_y";
+const MAX_X: &str = "max_x";
+const MAX_Y: &str = "max_y";
+const SPLIT_DIM: &str = "split_dim";
+const SPLIT_VALUE: &str = "split_value";
+const LEFT_CHILD: &str = "left_child";
+const RIGHT_CHILD: &str = "right_child";
+const FILE_ID: &str = "file_id";
+const ROW_OFFSET: &str = "row_offset";
+const NUM_ROWS: &str = "num_rows";
+
+/// Schema for inner node metadata
+pub fn inner_node_schema() -> Arc<Schema> {
+    Arc::new(Schema::new(vec![
+        Field::new(NODE_ID, DataType::UInt32, false),
+        Field::new(MIN_X, DataType::Float64, false),
+        Field::new(MIN_Y, DataType::Float64, false),
+        Field::new(MAX_X, DataType::Float64, false),
+        Field::new(MAX_Y, DataType::Float64, false),
+        Field::new(SPLIT_DIM, DataType::UInt8, false),
+        Field::new(SPLIT_VALUE, DataType::Float64, false),
+        Field::new(LEFT_CHILD, DataType::UInt32, false),
+        Field::new(RIGHT_CHILD, DataType::UInt32, false),
+    ]))
+}
+
+/// Schema for leaf node metadata
+pub fn leaf_node_schema() -> Arc<Schema> {
+    Arc::new(Schema::new(vec![
+        Field::new(NODE_ID, DataType::UInt32, false),
+        Field::new(MIN_X, DataType::Float64, false),
+        Field::new(MIN_Y, DataType::Float64, false),
+        Field::new(MAX_X, DataType::Float64, false),
+        Field::new(MAX_Y, DataType::Float64, false),
+        Field::new(FILE_ID, DataType::UInt32, false),
+        Field::new(ROW_OFFSET, DataType::UInt64, false),
+        Field::new(NUM_ROWS, DataType::UInt64, false),
+    ]))
+}
+
+/// BKD Tree node - either an inner node (with children) or a leaf node (with data location)
+#[derive(Debug, Clone, DeepSizeOf)]
+pub enum BKDNode {
+    Inner(BKDInnerNode),
+    Leaf(BKDLeafNode),
+}
+
+impl BKDNode {
+    pub fn bounds(&self) -> [f64; 4] {
+        match self {
+            BKDNode::Inner(inner) => inner.bounds,
+            BKDNode::Leaf(leaf) => leaf.bounds,
+        }
+    }
+
+    pub fn is_leaf(&self) -> bool {
+        matches!(self, BKDNode::Leaf(_))
+    }
+
+    pub fn as_inner(&self) -> Option<&BKDInnerNode> {
+        match self {
+            BKDNode::Inner(inner) => Some(inner),
+            _ => None,
+        }
+    }
+
+    pub fn as_leaf(&self) -> Option<&BKDLeafNode> {
+        match self {
+            BKDNode::Leaf(leaf) => Some(leaf),
+            _ => None,
+        }
+    }
+}
+
+/// Inner node in BKD tree - contains split information and child pointers
+#[derive(Debug, Clone, DeepSizeOf)]
+pub struct BKDInnerNode {
```

Comment:
I'm wondering if we can reduce the storage space occupied by bkd tree index by eliminating some fields. 

Since `split_dim` and `split_value` appear to be used only during index training and not during index searching, is serialization still necessary?

It seems possible to derive the child node ids based on the current node id, If using preorder traversal like `BKDTreeBuilder::build_recursive` to serialize a balanced binary tree, for example:
`left_child = node_id * 2 + 1` and `right_child = node_id * 2 + 2`